embedded/document/engine.go

```diff
@@ -757,3 +717,138 @@ func (e *Engine) readMetadataAndValue(key []byte, atTx uint64, skipIntegrityChec
 
 	return entry.Metadata(), v, nil
 }
+
+func (e *Engine) getDocumentID(doc *structpb.Struct, isInsert bool) (DocumentID, error) {
+	provisionedDocID, docIDProvisioned := doc.Fields[DocumentIDField]
+	var docID DocumentID
+	var err error
+	if docIDProvisioned {
+		docID, err = NewDocumentIDFromHexEncodedString(provisionedDocID.GetStringValue())
+		if err != nil {
+			return docID, err
+		}
+	} else {
+		if !isInsert {
+			return docID, fmt.Errorf("_id field is required when updating a document")
+		}
+
+		// generate document id
+		docID = NewDocumentIDFromTx(e.sqlEngine.GetStore().LastPrecommittedTxID())
+		doc.Fields[DocumentIDField] = structpb.NewStringValue(docID.EncodeToHexString())
+	}
+
+	return docID, nil
+}
+
+func (e *Engine) configureTxOptions(docIDProvisioned bool) *sql.TxOptions {
+	txOpts := sql.DefaultTxOptions()
+
+	if !docIDProvisioned {
+		// wait for indexing to include latest catalog changes
+		txOpts.
+			WithUnsafeMVCC(!docIDProvisioned).
+			WithSnapshotRenewalPeriod(0).
+			WithSnapshotMustIncludeTxID(func(lastPrecommittedTxID uint64) uint64 {
+				return e.sqlEngine.GetStore().MandatoryMVCCUpToTxID()
+			})
+	}
+	return txOpts
+}
+
+func (e *Engine) generateRowSpecForDocument(docID DocumentID, doc *structpb.Struct, table *sql.Table) (*sql.RowSpec, error) {
+	var cols []string
+	var values []sql.ValueExp
+
+	for _, col := range table.Cols() {
+		cols = append(cols, col.Name())
+
+		switch col.Name() {
+		case DocumentIDField:
+			// add document id to document
+			values = append(values, sql.NewBlob(docID[:]))
+		case DocumentBLOBField:
+			bs, err := json.Marshal(doc)
+			if err != nil {
+				return nil, err
+			}
+
+			values = append(values, sql.NewBlob(bs))
+		default:
+			if rval, ok := doc.Fields[col.Name()]; ok {
+				val, err := valueTypeToExp(col.Type(), rval)
+				if err != nil {
+					return nil, err
+				}
+				values = append(values, val)
+			} else {
+				values = append(values, &sql.NullValue{})
+			}
+		}
+	}
+
+	return sql.NewRowSpec(values), nil
+
+}
+
+func (e *Engine) BulkInsertDocuments(ctx context.Context, collectionName string, docs []*structpb.Struct) (docIDs []DocumentID, txID uint64, err error) {
```

Comment:
I'd expect `BulkInsertDocuments` or  `BulkUpdateDocuments` to use a generalized implementation of upsert method, and define single-document insertion or update in terms of bulk operations, then there is no duplicated code